### PR TITLE
[PLAT-366] Add download counts to admin

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

Admins want to see download counts, this will let them.

## Changes

- Add `get_all_downloads_on_date` so downloads can be queried on a certain date.
- JS and html additions to so Keen data can be viewed
- Event that sends daily download counts to Keen

## QA Notes

To set this up you must have a keen account and have your credentials set up with you `KEEN_CREDENTIALS` set to:
```
KEEN_CREDENTIALS = {
    'keen_ready': True
}
```
Also worth noting downloads numbers only include non-contributor downloads.

## Documentation

This will be demoed, so that's where the important info will be conferred. 

## Side Effects

The backdated data won't be included immediately, someone with prod access will have to combine the `benchmark.py` data with the preexisting `PageCounter` data.  

## Ticket

https://openscience.atlassian.net/projects/PLAT/issues/PLAT-366

### CODE REVIEW

File: osf/models/analytics.py [hunk 1]
```diff
@@ -63,6 +65,24 @@ class PageCounter(BaseModel):
...
+        :return: long sum:
+        """
+
+        # Get all PageCounters with data for the date made for all versions downloads,
+        # regex insures one colon so all versions are queried.
+        page_counters = PageCounter.objects.filter(date__has_key=date, _id__regex='^download:[^:]*:{1}[^:]*$')
+
+        # Get the total download numbers from the nested dict on the PageCounter by annotating it as daily_total then
+        # aggregating the sum.
+        sum = page_counters.annotate(daily_total=RawSQL("((date->%s->>'total')::int)", (date,))).aggregate(sum=Sum('daily_total'))['sum']
```
Comment (Johnetordoff) [inline osf/models/analytics.py]: Debated just changing the model instead of using this difficult query, but it works and keeping the numbers for a specific date is inherently difficult.

File: scripts/analytics/download_count_summary.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+from django.db.models import Q
+from website.app import init_app
+from scripts.analytics.base import SummaryAnalytics
+from osf.models import PageCounter
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+LOG_THRESHOLD = 11
```
Comment (erinspace) [inline scripts/analytics/download_count_summary.py]: Is this used anywhere?

File: scripts/analytics/download_count_summary.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+from dateutil.parser import parse
+from datetime import datetime, timedelta
+
+from django.db.models import Q
+from website.app import init_app
+from scripts.analytics.base import SummaryAnalytics
+from osf.models import PageCounter
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
```
Comment (erinspace) [inline scripts/analytics/download_count_summary.py]: Doesn't seem like anything is being logged in this summary, so this and the `logger` imports can probably be removed. Other unused imports include `requests` and `Q`

File: osf/models/analytics.py [hunk 4]
```diff
@@ -103,7 +123,7 @@ def update_counter(cls, page, node_info):
                     model_instance.date[date_string]['unique'] += 1
                 else:
                     # set the number of unique visitors to 1
-                    model_instance.date[date_string] = dict(unique=1)
+                    model_instance.date[date_string] = dict(unique=0)
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply osf/models/analytics.py]: There is no way to to test this currently because necessary waterbutler callback code with the user discerning logic isn't on prod. So this doesn't really do anything.

**Comment by erinspace** [comment]
Looks ok to me. Failing tests seem to be yarn related and/or are failing on `release/next-platform`, so just make sure to update when those things look OK to make sure other tests aren't unhappy

File: osf/models/analytics.py [hunk 5]
```diff
@@ -63,6 +65,26 @@ class PageCounter(BaseModel):
...
+    def get_all_downloads_on_date(date):
+        """
+        Queries the total number of downloads on a date
+        :param str date: must be formatted the same as a page counter key so 'yyyy/mm/dd'
+        :return: long sum:
+        """
+
+        # Get all PageCounters with data for the date made for all versions downloads,
+        # regex insures one colon so all versions are queried.
+        page_counters = PageCounter.objects.filter(date__has_key=date, _id__regex=PageCounter.DOWNLOAD_ALL_VERSIONS_ID_PATTERN)
```
Comment (sloria) [inline osf/models/analytics.py]: Seems like this should be a `classmethod` so that you don't have to hardcode `PageCounter`.

```python
page_counters = cls.objects.filter(...)
```

File: admin/static/js/pages/metrics-page.js [hunk 6]
```diff
@@ -50,5 +53,19 @@ keenAnalysis.ready(function() {
 
         Metrics.KeenRenderMetrics("#preprints-added", "line", preprint_created, 200);
     };
+    $('#downloads-range')[0].onclick = function() {
+        var download_count = new keenAnalysis.Query('sum', {
```
Comment (sloria) [inline admin/static/js/pages/metrics-page.js]: Minor: As per our style, use `downloadCount`.

File: osf_tests/test_analytics.py [hunk 7]
```diff
@@ -42,3 +44,125 @@ def test_increment_user_activity_counters(self):
...
+    page_counter_id = 'download:{}:{}'.format(project._id, file_node2.id)
+    page_counter, created = PageCounter.objects.select_for_update().get_or_create(_id=page_counter_id)
+    page_counter.date = {u'2018/02/04': {u'total': 4, u'unique': 26}}
+    page_counter.save()
+    return page_counter
+
+@pytest.fixture()
+def page_counter_for_individual_version(project, file_node3):
+    page_counter_id = 'download:{}:{}:0'.format(project._id, file_node3.id)
+    page_counter, created = PageCounter.objects.select_for_update().get_or_create(_id=page_counter_id)
```
Comment (sloria) [inline osf_tests/test_analytics.py]: Minor: No need to `select_for_update()` in tests.

File: osf_tests/test_analytics.py [hunk 8]
```diff
@@ -42,3 +44,125 @@ def test_increment_user_activity_counters(self):
...
+    page_counter.date = {u'2018/02/04': {u'total': 1, u'unique': 1}}
```
Comment (sloria) [inline osf_tests/test_analytics.py]: Minor: Why not pass the `date` to `get_or_create`? That would save a `save()`.

File: osf_tests/test_analytics.py [hunk 9]
```diff
@@ -42,3 +44,125 @@ def test_increment_user_activity_counters(self):
...
+@pytest.fixture()
+def page_counter(project, file_node):
+    page_counter_id = 'download:{}:{}'.format(project._id, file_node.id)
+    page_counter, created = PageCounter.objects.select_for_update().get_or_create(_id=page_counter_id)
+    page_counter.date = {u'2018/02/04': {u'total': 41, u'unique': 33}}
+    page_counter.save()
+    return page_counter
+
+@pytest.fixture()
+def page_counter2(project, file_node2):
```
Comment (sloria) [inline osf_tests/test_analytics.py]: Are these duplicative fixtures necessary? Can we test the behavior with just one fixture for each type of object?

File: osf_tests/test_analytics.py [hunk 10]
```diff
@@ -42,3 +44,125 @@ def test_increment_user_activity_counters(self):
...
+def user():
+    return UserFactory()
+
+
+@pytest.fixture()
+def project():
+    return ProjectFactory()
+
+@pytest.fixture()
+def project_with_contrib(user):
```
Comment (sloria) [inline osf_tests/test_analytics.py]: This should be unnecessary. Every project has a creator, so you can just use the `project` fixture and access `project.creator` on it.

File: osf_tests/test_analytics.py [hunk 11]
```diff
@@ -42,3 +44,125 @@ def test_increment_user_activity_counters(self):
...
+        total_downloads = PageCounter.get_all_downloads_on_date('2018/02/04')
+
+        assert total_downloads == 45
+
+    def test_get_all_downloads_on_date_exclude_versions(self, page_counter, page_counter2, page_counter_for_individual_version):
+        total_downloads = PageCounter.get_all_downloads_on_date('2018/02/04')
+
+        assert total_downloads == 45
+
+    def test_download_all_versions_regex(self):
```
Comment (sloria) [inline osf_tests/test_analytics.py]: This test doesn't actually need the database. So we can save a few seconds of database setup/teardown by explicitly marking the classes in this file that need a database and moving this one to its own class that doesn't get marked

```python
@pytest.mark.django_db
class TestAnalytics(..)

@pytest.mark.django_db
class TestPageCounter

class TestPageCounterRegex:
```

File: admin/static/js/metrics/metrics.es6.js [hunk 12]
```diff
@@ -1039,6 +1014,48 @@ var PreprintMetrics = function() {
...
+        filters: [{
+            property_name: 'action.type',
+            operator: 'eq',
+            property_value: 'download_file',
+            timezone: "UTC"
+        }]
+    });
+    renderKeenMetric("#number-of-unique-downloads", "metric", uniqueDownloadsQuery, defaultHeight, defaultColor, publicClient);
+
+    var download_count = new keenAnalysis.Query("sum", {
```
Comment (sloria) [inline admin/static/js/metrics/metrics.es6.js]: Minor: Should be `downloadCount`, as per our style guidelines.

File: osf_tests/test_analytics.py (hunk 9 above)
Comment (Johnetordoff) [reply osf_tests/test_analytics.py]: Yeah this does look confusing, but I want to show that we are summing over multiple PageCounter objects and each one corresponds with a file node. I'll try to clean it up and add comments. I'm not sure if this is a good test if there's isn't summing across multiple PageCounter objects.

File: osf/models/analytics.py [hunk 13]
```diff
@@ -102,8 +124,7 @@ def update_counter(cls, page, node_info):
                     # increment the number of unique visitors for today
                     model_instance.date[date_string]['unique'] += 1
                 else:
-                    # set the number of unique visitors to 1
-                    model_instance.date[date_string] = dict(unique=1)
+                    model_instance.date[date_string] = dict(unique=0)
```
Comment (sloria) [inline osf/models/analytics.py]: Why was this changed?

File: osf_tests/test_analytics.py [hunk 14]
```diff
@@ -42,3 +44,132 @@ def test_increment_user_activity_counters(self):
...
+        :param page_counter2: represents a page_counter for another file node being downloaded
+        """
+        date = datetime(2018, 2, 4)
+
+        total_downloads = PageCounter.get_all_downloads_on_date(date)
+
+        assert total_downloads == 45
+
+
+class TestPageCounterRegex:
```
Comment (sloria) [inline osf_tests/test_analytics.py]: These will still use a database because this whole file is marked with

```
pytestmark = pytest.mark.django_db
```

You'll need to explicitly mark the classes that need a db and keep this one unmarked.

File: scripts/analytics/download_count_summary.py [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+
+if __name__ == '__main__':
+    init_app()
+
+    download_count_summary = DownloadCountSummary()
+    args = download_count_summary.parse_args()
+    yesterday = args.yesterday
+    if yesterday:
+        date = (datetime.today() - timedelta(1)).date()
```
Comment (sloria) [inline scripts/analytics/download_count_summary.py]: Always use `timezone.now()` to get dates to ensure deterministic output (i.e. won't be different depending on location of server).

File: scripts/tests/test_downloads_summary.py [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+from scripts.analytics.download_count_summary import DownloadCountSummary
+
+
+@pytest.mark.django_db
+class TestDownloadCount(StorageTestCase):
+
+    def test_download_count(self):
+        # Keen does not allow same day requests so we have to do some time traveling to my birthday
+        timezone.now = mock.Mock(return_value=datetime.datetime(1991, 9, 25).replace(tzinfo=pytz.utc))
```
Comment (sloria) [inline scripts/tests/test_downloads_summary.py]: Use `mock.patch` to temporarily change the return values of things rather than assigning the attribute. That way, you don't have to worry about re-setting the attribute when the test is finished.

File: osf/models/analytics.py [hunk 17]
```diff
@@ -63,6 +65,26 @@ class PageCounter(BaseModel):
...
+        :return: long sum:
+        """
+        formatted_date = date.strftime('%Y/%m/%d')
+        # Get all PageCounters with data for the date made for all versions downloads,
+        # regex insures one colon so all versions are queried.
+        page_counters = cls.objects.filter(date__has_key=formatted_date, _id__regex=cls.DOWNLOAD_ALL_VERSIONS_ID_PATTERN)
+
+        # Get the total download numbers from the nested dict on the PageCounter by annotating it as daily_total then
+        # aggregating the sum.
+        daily_total = page_counters.annotate(daily_total=RawSQL("((date->%s->>'total')::int)", (formatted_date,))).aggregate(sum=Sum('daily_total'))['sum']
```
Comment (sloria) [inline osf/models/analytics.py]: You should be able to use `KeyTextTransform` here instead of `RawSQL`. Unfortunately, it's undocumented, but here's a thing: https://til.sloria.com/#2018-08-20-use-keytransform-to-query-on-jsonfields-in-django

File: osf_tests/test_analytics.py [hunk 18]
```diff
@@ -42,3 +44,132 @@ def test_increment_user_activity_counters(self):
...
+
+        match = re.match(PageCounter.DOWNLOAD_ALL_VERSIONS_ID_PATTERN, 'views:guid1:fileid')
+        assert not match
+
+        match = re.match(PageCounter.DOWNLOAD_ALL_VERSIONS_ID_PATTERN, 'download:guid1:fileid:0')
+        assert not match
+
+        match = re.match(PageCounter.DOWNLOAD_ALL_VERSIONS_ID_PATTERN, 'download:guid1:fileid')
+        assert match
+
```
Comment (sloria) [inline osf_tests/test_analytics.py]: Nit: Remove blank lines at ends of files.

This didn't get caught by flake8 because we're ignoring tests (for now).

File: osf/models/analytics.py (hunk 13 above)
Comment (Johnetordoff) [reply osf/models/analytics.py]: This was a pre-exsisting bug where the first unique user got double counted [here](https://github.com/CenterForOpenScience/osf.io/pull/8478/files/6eb96f148c82c8c8095dbfbd4a7b17fd7c178fe3#diff-4519792f13c44cf51a8eec0288fa674aR149) if IIRC.

File: osf/models/analytics.py (hunk 17 above)
Comment (Johnetordoff) [reply osf/models/analytics.py]: `KeyTextTransform` is very useful, but I think the RawSQL is necessary to cast the total as an int so it can be summed. With the query: `daily_total = page_counters.annotate(totals=KeyTextTransform(formatted_date, 'date')).aggregate(sum=Sum(KeyTextTransform('total', 'totals')))['sum']`

I'm getting the following:

```bash
 ProgrammingError: operator does not exist: text ->> unknown
 LINE 1: SELECT SUM(("totals" ->> 'total')) FROM (SELECT "osf_pagecou...
                              ^
 HINT:  No operator matches the given name and argument type(s). You might need to add explicit type casts.
```

Comment (sloria) [reply osf/models/analytics.py]: There's also a `KeyTransform` which I believe allows you to cast to any type.

Comment (Johnetordoff) [reply osf/models/analytics.py]: KeyTransform doesn't seem to do the trick, I'm getting the same problem. Looking at [the source](https://docs.djangoproject.com/en/2.1/_modules/django/contrib/postgres/fields/jsonb/) I can't see a obvious way of doing this with KeyTransform.

Comment (sloria) [reply osf/models/analytics.py]: Looking at the source of `Func` (from which `Transform` and `KeyTransform` inherit), it looks like you can pass `output_field`:

```python
KeyTransform('foo', 'bar', output_field=models.IntegerField())
```

Comment (Johnetordoff) [reply osf/models/analytics.py]: I've tried, the above and a few other approaches to casting this and they still bump into the same problem.

File: osf/models/analytics.py (hunk 13 above)
Comment (sloria) [reply osf/models/analytics.py]: The code you linked to is modifying `model_instance.unique`, not `model_instance.date`, so I'm unsure what you mean by the double-counting. Can you clarify?
